### ISSUE TITLE
IDLE: Add idlelib.config.IdleConf unittest

### ISSUE THREAD

**Issue by 9da18a29-a07d-4828-a801-24cc911246de** [python/cpython#75100]
BPO | [30917](https://bugs.python.org/issue30917)
--- | :---
Nosy | @terryjreedy, @vstinner, @mlouielu
PRs | <li>python/cpython#2691</li><li>python/cpython#2753</li><li>python/cpython#2754</li><li>python/cpython#2769</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/terryjreedy'
closed_at = <Date 2017-07-20.20:44:32.224>
created_at = <Date 2017-07-13.06:40:58.718>
labels = ['expert-IDLE', 'type-feature', '3.7']
title = 'IDLE: Add idlelib.config.IdleConf unittest'
updated_at = <Date 2017-07-20.20:44:32.192>
user = 'https://github.com/mlouielu'
```

bugs.python.org fields:
```python
activity = <Date 2017-07-20.20:44:32.192>
actor = 'terry.reedy'
assignee = 'terry.reedy'
closed = True
closed_date = <Date 2017-07-20.20:44:32.224>
closer = 'terry.reedy'
components = ['IDLE']
creation = <Date 2017-07-13.06:40:58.718>
creator = 'louielu'
dependencies = []
files = []
hgrepos = []
issue_num = 30917
keywords = []
message_count = 12.0
messages = ['298266', '298267', '298385', '298576', '298616', '298618', '298625', '298638', '298669', '298694', '298707', '298749']
nosy_count = 3.0
nosy_names = ['terry.reedy', 'vstinner', 'louielu']
pr_nums = ['2691', '2753', '2754', '2769']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue30917'
versions = ['Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by 9da18a29-a07d-4828-a801-24cc911246de** [comment]
Add idlelib.config.IdleConf unittest.

**Comment by 9da18a29-a07d-4828-a801-24cc911246de** [comment]
After this unittest is merged, I would like to first cleanup config.py source code, then modify it more general way.

**Comment by terryjreedy** [comment]
Yes, go ahead next with name changes and use of modern features such as sets and comprehensions.  Please discuss 'general changes before coding.  When we added Linux Modern keys a year ago, I combined two functions into IdleConf.current_colors_and_keys.  I suspect that there are other duplications to remove.  

Except for extensions, current values in default configs should not change, as doing so will break past versions.  So using them in tests is good, as it will detect unwanted changes to those files.  I hope the list of extensions goes down, but when it does, we can change the tests.

**Comment by terryjreedy** [comment]
Whatever is merged with PR_2691, further improvements can be made in further PRs either on this issue or others that work on config code.  

I still think this test is slower than it should be -- the change to deep copy either made no difference or made it slower, but I will try profiling before experimenting more.

In config, the warning prints in the directory search should use \_warn, which can be blocked and mocked while testing.

4% uncovered in the file is probably 6% of IdleConf.  At least some is error branches that might be interesting to trigger.

**Comment by terryjreedy** [comment]
New changeset f776eb0f0e046f2fa3a96540bb42d8cf970f6c55 by terryjreedy (Louie Lu) in branch 'master':
bpo-30917: IDLE: Add config.IdleConf unittests (bpo-2691)
https://github.com/python/cpython/commit/f776eb0f0e046f2fa3a96540bb42d8cf970f6c55

**Comment by terryjreedy** [comment]
The deepcopy does not work on 3.6.  Any fix should be applied to 3.7 as well.  If nothing else, we should try to reload with read_string.

This is a blocker for merging anything else into config and test_config.

**Comment by 9da18a29-a07d-4828-a801-24cc911246de** [comment]
[PR 2754](https://github.com/python/cpython/pull/2754) changed it to use `read_string`.

**Comment by terryjreedy** [comment]
Thanks.  I will look at it tomorrow and if it seems ok, prepare a new combined backport and see if it passes.

**Comment by vstinner** [comment]
Please see bpo-30968: "test_get_font (idlelib.idle_test.test_config.IdleConfTest) failure on x86 Windows7 3.x".

**Comment by terryjreedy** [comment]
New changeset ed014f7e135fe021837208960237d6c37afde5be by terryjreedy (Louie Lu) in branch 'master':
bpo-30917: IDLE: Fix mock_config deepcopy to read_string (bpo-2754)
https://github.com/python/cpython/commit/ed014f7e135fe021837208960237d6c37afde5be

**Comment by terryjreedy** [comment]
New changeset 65c24c846797b93d7adb72fc400f95a570f43fa9 by terryjreedy in branch '3.6':
[3.6] bpo-30917: IDLE: Add config.IdleConf unittests (GH-2691) (bpo-2753)
https://github.com/python/cpython/commit/65c24c846797b93d7adb72fc400f95a570f43fa9

**Comment by terryjreedy** [comment]
[PR2769](https://github.com/python/cpython/pull/2769) is the fix for master for test_get_font for bpo-30968.  The listing here is a cross-reference.  The fix was incorporated into the 3.6 backport [PR2753](https://github.com/python/cpython/pull/2753) before it was merged.

This issue is closed, having increased coverage from 46% to 96%.  Further work on config and test config should be new issues building on top of this one.
